obelisk/cpp/zoo/hardware/sensing/zed/zed2_sensors_interface.h

```diff
@@ -0,0 +1,420 @@
+
+            double timer_period = 1.0 / static_cast<double>(fps_);
+            auto timer = this->create_wall_timer(std::chrono::duration<double>(timer_period), timer_callback, cam_cbg_);
+            std::string timer_key = "timer_" + std::to_string(cam_index);
+            timers_[timer_key]    = timer;
+        }
+
+        // Create a timer for the publisher
+        auto pub_timer_callback = [this]() { this->publish_images(); };
+        if (pub_period_ == 0) {
+            pub_period_ = 1.0 / static_cast<double>(fps_);
+        }
+        auto pub_timer =
+            this->create_wall_timer(std::chrono::duration<double>(pub_period_), pub_timer_callback, cam_cbg_);
+        std::string pub_timer_key = "timer_pub";
+        timers_[pub_timer_key]    = pub_timer;
+
+        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+    }
+
+    /**
+     * @brief Callback for shutting down the node.
+     *
+     * @param prev_state Previous state of the lifecycle node.
+     * @return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+     * Callback return indicating success or failure.
+     */
+    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+    on_shutdown(const rclcpp_lifecycle::State& prev_state) {
+        obelisk::ObeliskSensor::on_shutdown(prev_state);
+        for (auto& cam : cams_) {
+            cam.second.close();
+        }
+        return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+    }
+
+  private:
+    // Key for the image publisher
+    std::string const pub_img_key_ = "pub_img";
+    double pub_period_;
+
+    // Callback group shared by all camera callbacks
+    rclcpp::CallbackGroup::SharedPtr cam_cbg_;
+
+    // Camera parameter maps
+    std::map<int, std::map<std::string, std::string>> cam_param_dicts_;
+    std::map<int, sl::Camera> cams_;
+    std::map<int, sl::Mat> cam_images_;
+    std::map<int, sl::RuntimeParameters> cam_rt_params_;
+    std::map<int, Eigen::Tensor<uint8_t, 3>> frames_;
+
+    // Shared camera parameters
+    sl::RESOLUTION resolution_;
+    int fps_;
+    bool depth_;
+    int height_;
+    int width_;
+
+    // Mutex for camera callback
+    std::mutex lock_;
+
+    /**
+     * @brief Set camera parameters from a YAML configuration file.
+     *
+     * @param params_path Path to the YAML configuration file.
+     */
+    void set_camera_params(const std::string& params_path) {
+        // Check if the file has a .yaml or .yml extension
+        if (params_path.substr(params_path.find_last_of(".") + 1) != "yaml" &&
+            params_path.substr(params_path.find_last_of(".") + 1) != "yml") {
+            std::string err_msg = "Invalid parameters file: " + params_path +
+                                  "! ObeliskZed2Sensors expects a YAML file with extension .yaml or .yml.";
+            RCLCPP_ERROR(this->get_logger(), err_msg.c_str());
```

Comment:
Once again with the extra string, probably won'r mark anymore if there are more.

Reply:
I'll resolve this conversation - let's consolidate to the other thread.